Add AnonymousParty superclass of Party

rnicoll opened the pull request:
Add AnonymousParty superclass of Party in preparation for anonymising parties stored in
contract states.

This is a redesign of the work for #100, although there are further PRs to complete the work. I'm trying to break it into more manageable chunks this time.

mikehearn commented on core/src/main/kotlin/net/corda/core/crypto/AnonymousParty.kt:
```diff
@@ -0,0 +1,36 @@
+import java.security.PublicKey
+
+/**
+ * The [Party] class represents an entity on the network, which is typically identified by a legal [name] and public key
```
The kdoc comment needs to be updated for the move, and Party needs to still have a kdoc comment as well.

mikehearn commented on core/src/main/kotlin/net/corda/core/node/services/IdentityService.kt:
```diff
@@ -23,3 +26,6 @@ interface IdentityService {
     fun partyFromName(name: String): Party?
 }
+
+fun IdentityService.deanonymiseParty(party: AnonymousParty) = partyFromKey(party.owningKey)
```
Why are these extension functions? Is the goal to give these functions better names? If so, let's just add the new names and deprecate the old ones.

rnicoll replied:
They're just nice helper methods to hide the referencing of fields inside the anonymous party. Mostly named differently because I wanted to avoid `partyFromParty()`

mikehearn replied:
Remember that extension functions aren't really ideal for Java users. You could also just put these in the interface itself, I think. Or add an extension to the Party class:

`fun AnonymousParty.deanonymise(service: IdentityService): Party`

That would probably make more sense.

rick-r3 replied:
Should this check if the party is a `Party` and do no work if so?

mikehearn commented on docs/source/key-concepts-core-types.rst:
```diff
@@ -87,12 +87,12 @@ Party and CompositeKey
+should be used, which contains a composite public key without any identifying information about who owns it.
 
-Identities of parties involved in signing a transaction can be represented simply by a ``CompositeKey``, or by further
+Identities of parties involved in signing a transaction can be represented simply by a ``AnonymousParty``, or by further
```
This second paragraph is now a little bit duplicative with the prior: you could just lead on with something like "If, on the other hand, a verified identity is required, use the Party class".

mikehearn commented on docs/source/release-notes.rst:
```diff
@@ -3,13 +3,12 @@ Release notes
 Here are brief summaries of what's changed between each snapshot release.
 
-Milestone 8
+Milestone 9
```
Eh? The section below this one is Milestone 7. Is the assumption that you'll rebase this onto the M8 release notes before merge?

rnicoll commented:
That should be ready for re-review @mikehearn

rick-r3 commented on core/src/main/kotlin/net/corda/core/crypto/AnonymousParty.kt:
```diff
@@ -0,0 +1,22 @@
+    constructor(owningKey: PublicKey) : this(owningKey.composite)
+
+    /** Anonymised parties do not include any detail apart from owning key, so equality is dependent solely on the key */
+    override fun equals(other: Any?): Boolean = other is Party && this.owningKey == other.owningKey
```
Is `is Party` here correct?  Surely should be `AnonymousParty` ?  With the current code no two `AnonymousParty` would ever be equal.  e.g. Can't use them in a HashMap.

rnicoll commented:
@rick-r3 I've changed the equality method (good catch), and had it check for the type being 'Party' and return as-is in that case.

I've also renamed the function to better fit the naming scheme used, and avoided an extension method based on Mike's feedback.

mikehearn commented:
Probably we should have a few unit tests that would have caught what Rick just spotted! Can you add some?

rnicoll commented:
Tests added

rnicoll commented:
Rebased to keep up to date with master. Could I get a review please @rick-r3 so I can get this merged before anything more changes? :)

rnicoll commented on core/src/main/kotlin/net/corda/core/crypto/Party.kt:
```diff
@@ -22,15 +20,8 @@ import java.security.PublicKey
-    /** Anonymised parties do not include any detail apart from owning key, so equality is dependent solely on the key */
-    override fun equals(other: Any?): Boolean = other is Party && this.owningKey == other.owningKey
-    override fun hashCode(): Int = owningKey.hashCode()
     override fun toString() = name
```
I think I've removed assumptions about what `toString()` produces, earlier. Have added a commit to make the results more predictable.
